## 2. Adam Optimizer and Dropout

In [19]:
import numpy as np
# Initial theta: [1, 1]
# Max iteration: 1000
iter_max = 1000

# SGD without momentum
def sgd(alpha):
    theta = np.array([1.0, 1.0])

    for i in range(iter_max):
        gradient = np.array([100 * theta[0], theta[1]])
        theta = theta - alpha * gradient

        if np.linalg.norm(gradient) < 0.001:
            return i + 1, theta

    return iter_max, theta

iter_count1, theta1 = sgd(0.1)
iter_count2, theta2 = sgd(0.005)

print(f'Alpha = 0.1, iteration count = {iter_count1}, theta = {theta1}')
print(f'Alpha = 0.005, iteration count = {iter_count2} theta = {theta2} ')

Alpha = 0.1, iteration count = 1000, theta = [           nan 1.74787125e-46]
Alpha = 0.005, iteration count = 1000 theta = [9.33263619e-302 6.65396858e-003] 


/var/folders/q4/9zg9sbjj6hxdsx18xy5j5ybw0000gn/T/ipykernel_44590/2637933898.py:11: RuntimeWarning: overflow encountered in scalar multiply
  gradient = np.array([100 * theta[0], theta[1]])
/var/folders/q4/9zg9sbjj6hxdsx18xy5j5ybw0000gn/T/ipykernel_44590/2637933898.py:12: RuntimeWarning: invalid value encountered in subtract
  theta = theta - alpha * gradient


In [ ]:
import numpy as np
# SGD with momentum
def sgd_momentum(alpha, beta):
    theta = np.array([1.0, 1.0])
    m = np.array([0.0, 0.0])

    for i in range(iter_max):
        gradient = np.array([100 * theta[0], theta[1]])
        # add the momentum in
        m = beta * m + (1 - beta) * gradient
        theta = theta - alpha * m

        if np.linalg.norm(gradient) < 0.001:
            return i + 1, theta
    
    return iter_max, theta 

iter_count3, theta3 = sgd_momentum(0.1, 0.9)
print(f'Alpha = 0.1, momentum = 0.9 iteration count = {iter_count3}, theta = {theta3}')

Alpha = 0.1, momentum = 0.9 iteration count = 145, theta = [0.0004848  0.00052577]


### 2.1 Result finds
Without momentum, for both alpha values, SGD wasn't able to converge to optimal solution.  
Both hit the max iteration counts at 1000.  
When alpha = 0.1, the learning rate is too large, and resulting a diverge making theta too big.
When alpha = 0.005, theta slowing converging but after 1000 iterations theta2 is still larger than  
the tolorance at 0.001

With momentum, theta went below 0.001 at 145 runs. So the momentum makes alpha = 0.1 work.

### 2.2 How momentum helps
m is the rolling average of the gradient of the past gradients. It sounds like we will take the next step  
based on the previous steps. "Momentum" sounds exactly like that. And this could prevent us from making  
huge jumps and possibly diverge. Instead, we could maintain a fairly low speed and wont't change directions  
while converging down.

### 2.3 Introducing v
The parameters with smaller gradients get larger updates and vise versa.  Such as theta2 = 1 gets bigger  
updates than theta1 = 100. This is due to the division. And larger value makes larger denominator making  
slower progress hence. This could help on even out between large and small parameters

### 2.4 Dropout
